torch/distributed/pipelining/stage.py

```diff
@@ -611,26 +610,66 @@ def backward_one_chunk(
+            else:
+                # perform the partial backwards for the inputs with a custom backward function
+                # when the "stage_ouput" is a loss, then it is a tensor, otherwise it is a tuple of tensors
+                if isinstance(bwd_kwargs["stage_output"], torch.Tensor):
+                    bwd_kwargs["stage_output"] = (bwd_kwargs["stage_output"],)
+
+                dinputs, param_groups = stage_backward_input(
+                    bwd_kwargs["stage_output"],
+                    bwd_kwargs["output_grads"],
+                    bwd_kwargs["input_values"],
+                    self.submod.parameters(),
+                )
+                # TODO: we dont need to save this, add to dw_runner?
+                self.backward_state[bwd_chunk_id] = (
+                    dinputs,
+                    input_values,
+                    param_groups,
+                    bwd_kwargs["stage_output"],
+                    bwd_kwargs["output_grads"],
+                )
+                self.grads_input = dinputs
+                # save a dw_runner with the `stage_backward_weight` function
+                self.dw_runner[bwd_chunk_id] = stage_backward_weight
         logger.debug("%s Backwarded chunk %s", self.log_prefix, bwd_chunk_id)
 
-        if not full_backward:
-            assert self.dw_builder, "Must provide dw_builder to run partial backward"
-            assert bwd_chunk_id not in self.dw_runner, (
-                f"{self.log_prefix} Attempted to run partial backward for chunk {bwd_chunk_id}"
-                " repeatedly without calling `backward_weight_one_chunk`"
-            )
-            dw_runner = self.dw_builder()
-            self.dw_runner[bwd_chunk_id] = dw_runner
-        elif self.dw_builder:
-            self.dw_builder()()
-
     def backward_weight_one_chunk(self, bwd_chunk_id: int):
-        assert bwd_chunk_id in self.dw_runner, (
-            f"{self.log_prefix} Attempted to run backward_weight_one_chunk for chunk {bwd_chunk_id}"
-            " without first calling `backward_one_chunk(full_backward=False)`"
-        )
-        self.dw_runner.pop(bwd_chunk_id)()
+        if self.dw_builder is not None:
+            assert bwd_chunk_id in self.dw_runner, (
+                f"{self.log_prefix} Attempted to run backward_weight_one_chunk for chunk {bwd_chunk_id}"
+                " without first calling `backward_one_chunk(full_backward=False)`"
+            )
+            self.dw_runner.pop(bwd_chunk_id)()
+        else:
+            (
+                dinputs,
+                input_values,
+                param_groups,
+                stage_output,
+                output_grads,
+            ) = self.backward_state.pop(bwd_chunk_id)
+            if self.stage_index != 0:
+                dweights = self.dw_runner.pop(bwd_chunk_id)(
+                    self.submod.parameters(), param_groups
+                )
+            else:
+                # TODO: figure out a better way to do this:
```

Comment:
i made a comment inside stage_backward_input, we might be able to fix this todo as well if we make the input fn do the intersect compute and then finish it off during W. seems ok to leave for follow up PR